### PR TITLE
feat: allow defining custom label schema for EntityPicker

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

This allows passing a custom schema to be used for entity picker options for example:

```yaml
entity:
  title: My Entity
  type: string
  ui:field: EntityPicker
  ui:options:
    allowArbitraryValues: false
    optionLabelSchema: ${{metadata.title}} (${{metadata.name}})
```

This does not work if `allowArbitraryValues` is set to `true`.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by drodil** [comment]
Would fix #15987

File: plugins/scaffolder/src/components/fields/EntityPicker/EntityPicker.tsx [hunk 1]
```diff
@@ -63,36 +75,60 @@ export const EntityPicker = (props: EntityPickerProps) => {
...
-        return humanizeEntityRef(
-          parseEntityRef(ref, { defaultKind, defaultNamespace }),
-          {
-            defaultKind,
-            defaultNamespace,
-          },
-        );
-      } catch (err) {
-        return ref;
+  const getOptionLabel = useCallback(
```
Comment (vinzscam) [inline plugins/scaffolder/src/components/fields/EntityPicker/EntityPicker.tsx:105]: just to flag that `useCallback` can be removed as `additionalFields` is not stable

**Comment by benjdlambert** [comment]
I'm seeing an increasing number of requests to do some form of templating in the form, and I want to make sure we standardize in a way that's not going to conflict with #16275.

#21754 is another one that's heading in a different direction, but similar use case.

Think it might be worth us setting up a BEP or something to decide how we want to solve both `nunjucks` templating and `templateStrings` being passed into the React components.

**Comment by mbenson** [comment]
> Think it might be worth us setting up a BEP or something to decide how we want to solve both nunjucks templating and templateStrings being passed into the React components.

How do we move forward on that?

**Comment by drodil** [comment]
@benjdlambert I agree that we need an agreement on how to use different templating parameters. Now I see there are 3 different ways used, `{{ value }}`, `${{ value }}`, and `@{{ value }}`. At least that should be the same for all use cases though these are separate issues from this.

**Comment by benjdlambert** [comment]
@mbenson @drodil are any of you willing to take the lead on drafting up a BEP that can cover some of this functionality?

`${{ }}` is currently reserved for the `nunjucks` templating, so I guess we need to work out that if we do `nunjucks` templating in the `parameters` section, how do we support passing in templating strings that are templated in the component in a nice way?

One option is to use something else that doesn't already clash, like the proposed `@{{ }}` or `{{ }}` but I think we just want to standardize on something so that we don't drift and have many different implementations of the same thing introducing confusion for integrators.

I wonder if there are other options so that we could support the same syntax, but being able to exclude the templating string somehow, but maybe that just gets to complicated.

**Comment by mbenson** [comment]
> @mbenson @drodil are any of you willing to take the lead on drafting up a BEP that can cover some of this functionality?
> 

I'm not _eager_ to commit to this larger process, but I do have a vested interest in the advancement of the platform, and could try to carve out some time for this in the near future if nobody else steps up (hopefully with more enthusiasm). I'd appreciate more info on the BEP process as relates to Backstage.

> `${{ }}` is currently reserved for the `nunjucks` templating, so I guess we need to work out that if we do `nunjucks` templating in the `parameters` section, how do we support passing in templating strings that are templated in the component in a nice way?
> 
> One option is to use something else that doesn't already clash, like the proposed `@{{ }}` or `{{ }}` but I think we just want to standardize on something so that we don't drift and have many different implementations of the same thing introducing confusion for integrators.
>

Point taken; I don't know how thoroughly I had really thought it out, but certainly my introduction of `@{ }}` was intended to avoid the clash; it's another matter whether I should have implemented the handling of said construct with nunjucks rather than lodash as I actually did. All these are things I had hoped to resolve in the course of a review and just wanted to open up discussion with a "pretty close" implementation.
 
> I wonder if there are other options so that we could support the same syntax, but being able to exclude the templating string somehow, but maybe that just gets to complicated.

I'm sure there would be ways to handle arbitrary objects that "block" the "normal" nunjucks handling, but like you I fear they could become convoluted.

**Comment by mbenson** [comment]
ping @benjdlambert

**Comment by mbenson** [comment]
WRT `@{{ }}`, I will note (however irrelevantly) that this choice was inspired by the feature of Apache Ant `macrodef` wherein such a dynamically defined element's attributes are referenced using `@{ }` constructs in contrast to the usual `${ }` property dereferencing syntax.

**Comment by drodil** [comment]
Not stale, needs discussion

**Comment by benjdlambert** [comment]
@mbenson thanks for the ping, sorry it's been a pretty hectic few weeks with BackstageCon / KubeCon and some PTO.

OK, I think that the additional effort of the BEP is probably worth it in this instance, I'm happy to draft something up which talks a little bit about all these areas so that we can settle on something and make a rough guide on how we're going to implement both https://github.com/backstage/backstage/issues/16275 and still support https://github.com/backstage/backstage/pull/23435 and this PR.

We don't have to implement it all at the same time either, but good to settle on a design and get some input and feedback from others too to make sure we've not missed anything.

Will get something up shortly!

**Comment by benjdlambert** [comment]
@mbenson I've started something just before the easter holidays - will keep adding to this ahead of next week: https://github.com/backstage/backstage/pull/23893

**Comment by asandoval** [comment]
➕ on the value of fixing this. Filter dropdown search on the side nav is unusable if the slug isn't human readable. It's a foreign key in my org, so it shouldn't be. But, the metadata.title is exactly what want to see displayed. This should provide us that.

**Comment by sirmmacias** [comment]
We are also interested on this, if some help is required let me know and we can try to help

**Comment by drodil** [comment]
What are the next steps to maybe have this merged? @benjdlambert ?

**Comment by benjdlambert** [comment]
@drodil I'm hopefully going to have time to look into the BEP next week as we've got a company hack week and should get some time outside of regular work to implement something.

Hopefully then that will unblock this work, along with other related issues. :pray:

**Comment by benjdlambert** [comment]
See some progress in #23893

**Comment by benjdlambert** [comment]
@drodil After chants in the Scaffolder SIG, we're wondering if it makes sense to push some of these options into the component itself and expose a few options that can be selected with `ui:options` rather than exposing the templating string as such.

```yaml
ui:options:
  labelTemplate: titleAndName
```

Think that's probably one way forward?

**Comment by drodil** [comment]
@benjdlambert that's better than using the `EntityPresentationApi` which crashes with a large catalog. See #25340. This basically prevents us from updating to 1.28.2 for now.

**Pull request closed without merging**